models/evc.py

```diff
@@ -400,6 +422,56 @@ def archive(self):
         """Archive this EVC on deletion."""
         self.archived = True
 
+    def use_available_uni_tags(self):
+        """Use UNI tags if available, raise ValueError otherwise
+        This method is used when creating an EVC
+
+        Note:
+            ValueError is not expected to hit because self._validate()
+            ensures uni.tag availability already.
+        """
+        uni_a = self.uni_a
+        try:
+            self._use_uni_vlan(uni_a)
+        except ValueError as err:
+            raise err
+        uni_z = self.uni_z
+        try:
+            self._use_uni_vlan(uni_z)
+        except ValueError as err:
+            self.make_uni_vlan_available(uni_a)
+            raise err
+
+    def _use_uni_vlan(self, uni: UNI):
+        """Use vlan from uni"""
+        if uni.user_tag is None:
+            return
+        tag = uni.user_tag.value
+        if isinstance(tag, int):
+            success = uni.interface.use_tag(uni.user_tag)
+            if not success:
+                intf = uni.interface.id
+                raise ValueError(f"TAG {tag} is not available in {intf}")
+            notify_uni_available_tags(self._controller, {"uni": uni})
+
+    def make_uni_vlan_available(self, uni: UNI):
+        """Make UNI vlan available"""
+        if uni.user_tag is None:
+            return
+        tag = uni.user_tag.value
+        if isinstance(tag, int):
+            success = uni.interface.make_tag_available(uni.user_tag)
+            if not success:
+                intf = uni.interface.id
+                log.warning(f"Tag {tag} was already available in {intf}")
```

Comment:
Ideally this should be logged as `error`, since it's an unexpected conflict

Reply:
Changed